cpp/include/cudf/strings/detail/copy_if_else.cuh

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/valid_if.cuh>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+
+namespace cudf
+{
+namespace strings
+{
+namespace detail
+{
+
+/**
+ * @brief Returns a new strings column using the specified Filter to select
+ * strings from the lhs column or the rhs column.
+ *
+ * ```
+ * output[i] = filter_fn(i) ? lhs(i) : rhs(i)
+ * ```
+ *
+ * @tparam Filter Functor that takes an index and returns a boolean.
+ *
+ * @param lhs Strings instance to copy if Filter is true
+ * @param rhs Strings instance to copy if Filter is false
+ * @param mr Resource for allocating device memory.
+ * @param stream CUDA stream to use kernels in this method.
+ * @return New strings column.
+ */
+template<typename Filter>
+std::unique_ptr<cudf::column> copy_if_else( strings_column_view const& lhs,
+                                            strings_column_view const& rhs,
+                                            Filter filter_fn,
+                                            rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource(),
+                                            cudaStream_t stream=0 )
+{
+    CUDF_EXPECTS(lhs.size() == rhs.size(), "Both columns must be the same size");
+    auto strings_count = lhs.size();
+    if( strings_count == 0 )
+        return make_empty_strings_column(mr,stream);
+
+    auto execpol = rmm::exec_policy(stream);
+    auto lhs_column = column_device_view::create(lhs.parent(),stream);
+    auto d_lhs = *lhs_column;
+    auto rhs_column = column_device_view::create(rhs.parent(),stream);
+    auto d_rhs = *rhs_column;
+    // create null mask
+    rmm::device_buffer null_mask;
+    size_type null_count = 0;
+    if( lhs.has_nulls() || rhs.has_nulls() )
+    {
+        // Should we make the caller do this?
```

Comment:
Do you usually make the caller compute the valid mask to avoid duplicate code?

Reply:
In case of gather caller computes the null mask, is it same in case of `copy_if_else`?

Reply:
The caller setting the null-mask was a pattern in a couple of previous `strings::detail` functions created for general cudf. It does not appear appropriate in this case. I removed the comment.

Reply:
Why is it not appropriate in this case? I think we are trying to standardize on that pattern.

Reply:
In the copy-if-else code for fixed-types has a single kernel that includes the null-mask resolution logic in it. So I thought the pattern would not apply here since the null-mask code logic is not independent.

Reply:
Ah, OK.

Reply:
The inconsistency worries me. But not enough to stall this.